File: integrated-project/mindboost-javafx/src/main/java/com/mindboost/app/service/AdminStatisticsService.java
package com.mindboost.app.service;

import com.mindboost.app.config.DatabaseManager;
import com.mindboost.app.model.GeneralTest;
import com.mindboost.app.model.Score;
import com.mindboost.app.model.SpecificScore;
import com.mindboost.app.model.SpecificTest;
import jakarta.persistence.EntityManager;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AdminStatisticsService {
    private final DatabaseManager databaseManager;

    public AdminStatisticsService(DatabaseManager databaseManager) {
        this.databaseManager = databaseManager;
    }

    public AdminStats getDashboardData() {
        try (EntityManager em = databaseManager.createEntityManager()) {
            List<Score> generalScores = em.createQuery("from Score", Score.class).getResultList();
            List<SpecificScore> specificScores = em.createQuery("from SpecificScore", SpecificScore.class).getResultList();
            Map<Integer, GeneralTest> generalTests = indexById(em.createQuery("from GeneralTest", GeneralTest.class).getResultList());
            Map<Integer, SpecificTest> specificTests = indexById(em.createQuery("from SpecificTest", SpecificTest.class).getResultList());

            Map<Integer, Boolean> studentIds = new HashMap<>();
            Map<String, Integer> categoryCounts = new HashMap<>();
            Map<String, Integer> levelCounts = new HashMap<>();
            Map<String, Integer> scoreRanges = new HashMap<>();
            scoreRanges.put("0-24", 0);
            scoreRanges.put("25-49", 0);
            scoreRanges.put("50-74", 0);
            scoreRanges.put("75-100", 0);
            Map<String, Integer> weeklyCounts = new HashMap<>();
            Map<String, Integer> generalUsage = new HashMap<>();
            Map<String, Integer> specificUsage = new HashMap<>();

            int generalSum = 0;
            int generalCount = 0;

            for (Score score : generalScores) {
                studentIds.put(score.getUserId(), true);
                generalSum += score.getPercentage() == null ? 0 : score.getPercentage();
                generalCount++;
                String title = generalTests.get(score.getGeneralTestId()) != null
                    ? generalTests.get(score.getGeneralTestId()).getTitle()
                    : "Test général";
                generalUsage.put(title, generalUsage.getOrDefault(title, 0) + 1);
            }

            for (SpecificScore score : specificScores) {
                studentIds.put(score.getUserId(), true);
                String category = score.getCategory() == null ? "Non définie" : score.getCategory();
                String level = score.getLevel() == null ? "Non défini" : score.getLevel();
                int percentage = score.getPercentage() == null ? 0 : score.getPercentage();

                categoryCounts.put(category, categoryCounts.getOrDefault(category, 0) + 1);
                levelCounts.put(level, levelCounts.getOrDefault(level, 0) + 1);

                if (percentage <= 24) {
                    scoreRanges.put("0-24", scoreRanges.get("0-24") + 1);
                } else if (percentage <= 49) {
                    scoreRanges.put("25-49", scoreRanges.get("25-49") + 1);
                } else if (percentage <= 74) {
                    scoreRanges.put("50-74", scoreRanges.get("50-74") + 1);
                } else {
                    scoreRanges.put("75-100", scoreRanges.get("75-100") + 1);
                }

                String weekLabel = "Semaine " + (score.getWeekNumber() == null ? 0 : score.getWeekNumber());
                weeklyCounts.put(weekLabel, weeklyCounts.getOrDefault(weekLabel, 0) + 1);

                String title = specificTests.get(score.getSpecificTestId()) != null
                    ? specificTests.get(score.getSpecificTestId()).getTitle()
                    : "Test spécifique";
                specificUsage.put(title, specificUsage.getOrDefault(title, 0) + 1);
            }

            return new AdminStats(
                generalScores.size(),
                specificScores.size(),
                studentIds.size(),
                generalCount > 0 ? Math.round((float) generalSum / generalCount) : 0,
                categoryCounts,
                levelCounts,
                scoreRanges,
                weeklyCounts,
                generalUsage,
                specificUsage
            );
        }
    }

    private <T> Map<Integer, T> indexById(List<T> entities) {
        Map<Integer, T> indexed = new HashMap<>();
        for (T entity : entities) {
            try {
                Integer id = (Integer) entity.getClass().getMethod("getId").invoke(entity);
                if (id != null) {
                    indexed.put(id, entity);
                }
            } catch (Exception ignored) {
            }
        }
        return indexed;
    }

    public record AdminStats(
        int totalGeneralResults,
        int totalSpecificResults,
        int totalStudentsTested,
        int averageGeneralScore,
        Map<String, Integer> categoryCounts,
        Map<String, Integer> levelCounts,
        Map<String, Integer> scoreRanges,
        Map<String, Integer> weeklyCounts,
        Map<String, Integer> generalUsage,
        Map<String, Integer> specificUsage
    ) {}
}

File: symfony/src/Service/AdminStatisticsService.php
<?php

namespace App\Service;

use App\Entity\GeneralTest;
use App\Entity\Score;
use App\Entity\SpecificScore;
use App\Entity\SpecificTest;
use Doctrine\ORM\EntityManagerInterface;

class AdminStatisticsService
{
    public function __construct(private EntityManagerInterface $entityManager)
    {
    }

    public function getDashboardData(): array
    {
        $generalScores = $this->entityManager->getRepository(Score::class)->findAll();
        $specificScores = $this->entityManager->getRepository(SpecificScore::class)->findAll();
        $generalTests = $this->indexById($this->entityManager->getRepository(GeneralTest::class)->findAll());
        $specificTests = $this->indexById($this->entityManager->getRepository(SpecificTest::class)->findAll());

        $studentIds = [];
        $categoryCounts = [];
        $levelCounts = [];
        $scoreRanges = [
            '0-24' => 0,
            '25-49' => 0,
            '50-74' => 0,
            '75-100' => 0,
        ];
        $weeklyCounts = [];
        $generalUsage = [];
        $specificUsage = [];

        $generalSum = 0;
        $generalCount = 0;

        foreach ($generalScores as $score) {
            $studentIds[$score->getUserId()] = true;
            $generalSum += (int) ($score->getPercentage() ?? 0);
            $generalCount++;

            $title = $generalTests[$score->getGeneralTestId()]?->getTitle() ?? 'Test général';
            $generalUsage[$title] = ($generalUsage[$title] ?? 0) + 1;
        }

        foreach ($specificScores as $score) {
            $studentIds[$score->getUserId()] = true;

            $category = $score->getCategory() ?: 'Non définie';
            $level = $score->getLevel() ?: 'Non défini';
            $percentage = (int) ($score->getPercentage() ?? 0);

            $categoryCounts[$category] = ($categoryCounts[$category] ?? 0) + 1;
            $levelCounts[$level] = ($levelCounts[$level] ?? 0) + 1;

            if ($percentage <= 24) {
                $scoreRanges['0-24']++;
            } elseif ($percentage <= 49) {
                $scoreRanges['25-49']++;
            } elseif ($percentage <= 74) {
                $scoreRanges['50-74']++;
            } else {
                $scoreRanges['75-100']++;
            }

            $weekLabel = 'Semaine ' . ($score->getWeekNumber() ?? 0);
            $weeklyCounts[$weekLabel] = ($weeklyCounts[$weekLabel] ?? 0) + 1;

            $title = $specificTests[$score->getSpecificTestId()]?->getTitle() ?? 'Test spécifique';
            $specificUsage[$title] = ($specificUsage[$title] ?? 0) + 1;
        }

        ksort($weeklyCounts);
        arsort($categoryCounts);
        arsort($levelCounts);
        arsort($generalUsage);
        arsort($specificUsage);

        return [
            'kpis' => [
                'total_general_results' => count($generalScores),
                'total_specific_results' => count($specificScores),
                'total_students_tested' => count($studentIds),
                'average_general_score' => $generalCount > 0 ? round($generalSum / $generalCount) : 0,
            ],

            'category_counts' => $categoryCounts,
            'level_counts' => $levelCounts,
            'score_ranges' => $scoreRanges,
            'weekly_counts' => $weeklyCounts,
            'general_usage' => $generalUsage,
            'specific_usage' => $specificUsage,

            'category_labels' => array_keys($categoryCounts),
            'category_values' => array_values($categoryCounts),

            'level_labels' => array_keys($levelCounts),
            'level_values' => array_values($levelCounts),

            'score_range_labels' => array_keys($scoreRanges),
            'score_range_values' => array_values($scoreRanges),

            'weekly_labels' => array_keys($weeklyCounts),
            'weekly_values' => array_values($weeklyCounts),

            'general_usage_labels' => array_slice(array_keys($generalUsage), 0, 8),
            'general_usage_values' => array_slice(array_values($generalUsage), 0, 8),

            'specific_usage_labels' => array_slice(array_keys($specificUsage), 0, 8),
            'specific_usage_values' => array_slice(array_values($specificUsage), 0, 8),
        ];
    }

    private function indexById(array $entities): array
    {
        $indexed = [];

        foreach ($entities as $entity) {
            if (method_exists($entity, 'getId')) {
                $indexed[$entity->getId()] = $entity;
            }
        }

        return $indexed;
    }
}